micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java

```diff
@@ -104,4 +104,14 @@ protected DistributionStatisticConfig defaultHistogramConfig() {
             .merge(DistributionStatisticConfig.DEFAULT);
     }
 
+    @Override
+    public void close() {
+        stop();
+        getMeters().stream()
+            .filter(meter -> meter instanceof StepMeter)
+            .map(meter -> (StepMeter) meter)
+            .forEach(StepMeter::_manualRollover);
```

Comment:
@shakuzen  I do have a concern with this particular close() method.

Scenario:
Assume that the step is 10 seconds and the application's graceful shutdown is initiated exactly or around 10th seconds. So, the scheduler would have already initiated the pushiSafely() and there are chances that close() is called around the same time. For a step meter, I believe there might be a scenario where manualRollover() will call sumThenReset() of StepValue and the pubishSafely initiated by Scheduler might call the rollCount(). This will again trigger a sumThenReset() which will make the data disappear or rest to zero.

Though this might be an edge case, registries with shorter step have more probability of running into this particular issue.

Reply:
Thanks for taking a look and giving feedback so quickly. I think the general scenario is summarized as the registry being closed during a scheduled publish. It does sound like some data can be lost then. Would you mind opening an issue for this so we don't forget? I'm not immediately sure what we'll want to do to mitigate this. I can take a closer look tomorrow. We should be able to write a test case that simulates this scenario.